### PR TITLE
Exposing web3.eth.Contract.setProvider() as per public documentation

### LINKED ISSUE #4770: setProvider "does not exist on type 'typeof import(..." in typescript.
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Can't call Web3EthContract.setProvider(ethereum)

function doesn't exist in the index.d.ts 

```
/*
    This file is part of web3.js.
    web3.js is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.
    web3.js is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Lesser General Public License for more details.
    You should have received a copy of the GNU Lesser General Public License
    along with web3.js.  If not, see <http://www.gnu.org/licenses/>.
*/
/**
 * @file index.d.ts
 * @author Josh Stevens <joshstevens19@hotmail.co.uk>
 * @date 2018
 */

import BN = require('bn.js');
import {Common, PromiEvent, provider, hardfork, chain, BlockNumber, PastLogsOptions, LogsOptions} from 'web3-core';
import {AbiItem} from 'web3-utils';

// TODO: Add generic type!
export class Contract {
    constructor(
        jsonInterface: AbiItem[],
        address?: string,
        options?: ContractOptions
    );

    private _address: string;
    private _jsonInterface: AbiItem[];
    defaultAccount: string | null;
    defaultBlock: BlockNumber;
    defaultCommon: Common;
    defaultHardfork: hardfork;
    defaultChain: chain;
    transactionPollingTimeout: number;
    transactionConfirmationBlocks: number;
    transactionBlockTimeout: number;
    handleRevert: boolean;

    options: Options;

    clone(): Contract;

    deploy(options: DeployOptions): ContractSendMethod;

    methods: any;

    once(
        event: string,
        callback: (error: Error, event: EventData) => void
    ): void;
    once(
        event: string,
        options: EventOptions,
        callback: (error: Error, event: EventData) => void
    ): void;

    events: any;

    getPastEvents(event: string): Promise<EventData[]>;
    getPastEvents(
        event: string,
        options: PastEventOptions,
        callback: (error: Error, event: EventData) => void
    ): Promise<EventData[]>;
    getPastEvents(event: string, options: PastEventOptions): Promise<EventData[]>;
    getPastEvents(
        event: string,
        callback: (error: Error, event: EventData) => void
    ): Promise<EventData[]>;
}

export interface Options extends ContractOptions {
    address: string;
    jsonInterface: AbiItem[];
}

export interface DeployOptions {
    data: string;
    arguments?: any[];
}

export interface ContractSendMethod {
    send(
        options: SendOptions,
        callback?: (err: Error, transactionHash: string) => void
    ): PromiEvent<Contract>;

    call(
        options?: CallOptions,
        callback?: (err: Error, result: any) => void
    ): Promise<any>;

    estimateGas(
        options: EstimateGasOptions,
        callback?: (err: Error, gas: number) => void
    ): Promise<number>;

    estimateGas(callback: (err: Error, gas: number) => void): Promise<number>;

    estimateGas(
        options: EstimateGasOptions,
        callback: (err: Error, gas: number) => void
    ): Promise<number>;

    estimateGas(options: EstimateGasOptions): Promise<number>;

    estimateGas(): Promise<number>;

    encodeABI(): string;
}

export interface CallOptions {
    from?: string;
    gasPrice?: string;
    gas?: number;
}

export interface SendOptions {
    from: string;
    gasPrice?: string;
    gas?: number;
    value?: number | string | BN;
    nonce?: number;
}

export interface EstimateGasOptions {
    from?: string;
    gas?: number;
    value?: number | string | BN;
}

export interface ContractOptions {
    // Sender to use for contract calls
    from?: string;
    // Gas price to use for contract calls
    gasPrice?: string;
    // Gas to use for contract calls
    gas?: number;
    // Contract code
    data?: string;
}

export interface PastEventOptions extends PastLogsOptions {
    filter?: Filter;
}

export interface EventOptions extends LogsOptions {
    filter?: Filter;
}

export interface Filter {
    [key: string]: number | string | string[] | number[];
}

export interface EventData {
    returnValues: {
        [key: string]: any;
    };
    raw: {
        data: string;
        topics: string[];
    };
    event: string;
    signature: string;
    logIndex: number;
    transactionIndex: number;
    transactionHash: string;
    blockHash: string;
    blockNumber: number;
    address: string;
}

```

### Expected Behavior

I want to use setProvider

### Steps to Reproduce

```

import Web3 from 'web3'
import Web3EthContract from 'web3-eth-contract'


const { ethereum } = window as any
    const metamaskIsInstalled = ethereum && ethereum.isMetaMask
    if (metamaskIsInstalled) {
      Web3EthContract.setProvider(ethereum)
      const web3 = new Web3(ethereum)

     ....


```

### Web3.js Version

1.7.0

### Environment

- Operating System: 
- Browser: 
- Node.js Version: 
- NPM Version: 

### Anything Else?

_No response_

### PR DESCRIPTION
**Pull request by sdesalas**
## Description

Fixes #4770 `setProvider "does not exist on type 'typeof import(..." in typescript.`

## Type of change

- [x] `*.ts` Type Definitions 

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas (_**No code, only a type definition for TS**_).
- [x] I have made corresponding changes to the documentation (_**Fixes type defs in line with public documentation**_).
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules (_**No dependent changes**_).
- [x] I ran `npm run dtslint` with success and extended the tests and types if necessary.
- [x] I ran `npm run test:unit` with success.
- [x] I ran `npm run test:cov` and my test cases cover all the lines and branches of the added code.
- [x] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [x] I have tested my code on the live network (_**Unnecessary, only a missing type definition for TS**_).
- [x] I have checked the Deploy Preview and it looks correct (_**Unnecessary, only a missing type definition for TS**_).
- [x] I have updated the `CHANGELOG.md` file in the root folder (_**Not a notable change**_).

### CODE REVIEW

File: packages/web3-eth-contract/types/index.d.ts [hunk 1]
```diff
@@ -29,6 +29,8 @@ export class Contract {
         options?: ContractOptions
     );
 
+    static setProvider(provider: string): void;
```
Comment (nazarhussain) [inline packages/web3-eth-contract/types/index.d.ts]: The `setProvider` accepts an additional parameter for accounts as well. 

https://github.com/ChainSafe/web3.js/blob/832271a38e555fa62c843a05aba4eb5514f53d8a/packages/web3-eth-contract/src/index.js#L338

**Review by nazarhussain — CHANGES_REQUESTED** [review]

File: packages/web3-eth-contract/types/index.d.ts (hunk 1 above)
Comment (sdesalas) [reply packages/web3-eth-contract/types/index.d.ts]: Hiya, thanks good point. Added TS defs for optional `accounts` parameter as suggested.

Comment (sdesalas) [reply packages/web3-eth-contract/types/index.d.ts]: And parameter `provider` is of type [`provider`](https://github.com/ChainSafe/web3.js/blob/0b890b70b977451836dc693614960256bcbcb6d8/packages/web3-core/types/index.d.ts#L451) too (from `web3-core`). Fixing.

**Review by avkos — APPROVED** [review]

**Comment by sdesalas** [comment]
Hiya. Can someone rerun the flaky `e2e_windows` tests? Thanks!

**Review by nazarhussain — APPROVED** [review]

**Comment by nazarhussain** [comment]
@sdesalas The `e2e_windows` tests requires to run from core contributors. Soon we will merge your PR in one of our branches and then run these tests.

**Comment by sdesalas** [comment]
👍 Sure

**Review by jdevcs — APPROVED** [review]

**Comment by sdesalas** [comment]
Updating latest changes from upstream.

**Comment by sdesalas** [comment]
Wohooo! 🙌

**Pull request merged**